master.js

```diff
@@ -339,37 +380,92 @@ class ControlConnection extends BaseConnection {
 			});
 		}
 		return { list };
 	}
 
 	// XXX should probably add a hook for slave reuqests?
-	async createInstanceCommandRequestHandler(message) {
-		let { slave_id, name } = message.data;
-		let connection = slaveConnections.get(slave_id);
-		if (!connection) {
-			throw new errors.RequestError("Slave is not connected");
-		}
-
-		await link.messages.createInstance.send(connection, {
-			id: Math.random() * 2**31 | 0, // TODO: add id option
-			options: {
-				'name': name,
-				'description': config.description,
-				'visibility': config.visibility,
-				'username': config.username,
-				'token': config.token,
-				'game_password': config.game_password,
-				'verify_user_identity': config.verify_user_identity,
-				'allow_commands': config.allow_commands,
-				'auto_pause': config.auto_pause,
+	async createInstanceRequestHandler(message) {
+		let instanceConfig = new config.InstanceConfig();
+		await instanceConfig.load(message.data.serialized_config);
+
+		let instanceId = instanceConfig.get("instance.id")
+		if (db.instances.has(instanceId)) {
+			throw new errors.RequestError(`Instance with ID ${instanceId} already exists`);
+		}
+		db.instances.set(instanceId, instanceConfig);
+	}
+
+	async getInstanceConfigRequestHandler(message) {
+		let instanceConfig = db.instances.get(message.data.instance_id);
+		if (!instanceConfig) {
+			throw new errors.RequestError(`Instance with ID ${message.data.instance_id} does not exist`);
+		}
+
+		return {
+			serialized_config: instanceConfig.serialize(),
+		}
+	}
+
+	async setInstanceConfigFieldRequestHandler(message) {
+		let instanceConfig = db.instances.get(message.data.instance_id);
+		if (!instanceConfig) {
+			throw new errors.RequestError(`Instance with ID ${message.data.instance_id} does not exist`);
+		}
+
+		if (message.data.field === "instance.assigned_slave") {
+			throw new errors.RequestError("instance.assigned_slave must be set through the assign-slave interface");
+		}
+
+		if (message.data.field === "instance.id") {
+			// XXX is this worth implementing?  It's race condition galore.
+			throw new errors.RequestError("Setting instance.id is not supported");
+		}
+
+		instanceConfig.set(message.data.field, message.data.value);
+
+		// Push updated config to slave
+		let slaveId = instanceConfig.get('instance.assigned_slave');
+		if (slaveId) {
+			let connection = slaveConnections.get(slaveId);
+			if (connection) {
+				await link.messages.assignInstance.send(connection, {
+					instance_id: instanceConfig.get("instance.id"),
+					serialized_config: instanceConfig.serialize(),
+				});
 			}
+		}
+	}
+
+	async assignInstanceCommandRequestHandler(message, request) {
+		let instanceConfig = db.instances.get(message.data.instance_id);
+		if (!instanceConfig) {
+			throw new errors.RequestError(`Instance with ID ${message.data.instance_id} does not exist`);
+		}
+
+		// XXX: Should the instance be stopped if it's running on another slave?
```

Comment:
For now, I would say that the user has to stop the instance himself and handle the migration. We might think about developing migration later, cause it's a complexity multiplier. Don't want to feature-creep at this moment.